lib/user_interface.rb

```diff
@@ -91,28 +47,76 @@ def display_letter_header(letter)
   end
 
   def continue
-    output.print CONTINUE_MESSAGE
+    output.print Constants::CONTINUE_MESSAGE
     input.getch
   end
 
   def search_term
-    output.print SEARCH_MESSAGE
+    output.print Constants::SEARCH_MESSAGE
     collect_vaild_input { |user_input| validator.valid_string?(user_input) }
   end
 
   def search_again?
-    output.print ANOTHER_SEARCH_PROMPT
-    collect_vaild_input { |user_input| validator.valid_yes_no_answer?(user_input) }.downcase == YES_REPLY
+    boolen_choice?(Constants::ANOTHER_SEARCH_PROMPT)
+  end
+
+  def choose_contact(contacts)
+    display_all_with_index(contacts)
+    ask_for_index(contacts.length).to_i
+  end
+
+  def edit_field(contact)
+    display_contact(contact)
+    output.print Constants::FIELD_CHOICE_PROMPT
+
+    field = collect_field_key
+    output.print Constants::FIELDS_TO_PROMPTS[field]
+    value = collect_valid_field(field)
+    { field => value }
+  end
+
+  def update_another_field?
+    boolen_choice?(Constants::ANOTHER_EDIT_PROMPT)
+  end
+
+  def update_another_contact?
+    boolen_choice?(Constants::ANOTHER_UPDATE_PROMPT)
   end
 
   private
 
+  def ask_for_index(array_length)
+    output.print Constants::CONTACT_INDEX_PROMPT
+    collect_vaild_input { |user_input| validator.valid_index?(user_input, array_length) }
+  end
+
+  def display_all_with_index(contacts)
+    contacts.each_with_index do |contact, index|
+      output.print "[#{index}]"
+      display_contact(contact)
+    end
+  end
+
+  def collect_field_key
+    field_name = collect_vaild_input { |user_input| validator.valid_field_name?(user_input) }
+    Constants::FIELDS_TO_DISPLAY_NAMES.key(field_name.capitalize + ': ')
+  end
+
+  def collect_valid_field(field)
```

Comment:
There is something a bit confusing going on with the naming here:

`collect_field_key` calls `valid_field_name`
 
`collect_valid_field` calls `valid_field`


What do you think about renaming these four methods so that they are easier to decipher:

`collect_field_name` calls `valid_field_name` 

`collect_field_value` calls `valid_field_value`


field name, field name.
field value, field value.
all super clear. 

Call the same concepts with the same names. 
That will make things more readable.